### ISSUE TITLE
Streaming http responses are delayed until more chunks arrive

### ISSUE THREAD

**Issue by FelixZY** [oven-sh/bun#13696]
### What version of Bun is running?

1.1.26+0a37423ba

### What platform is your computer?

Linux 5.15.0-119-generic x86_64 x86_64

### What steps can reproduce the bug?

1. Create a new project following [the official getting started guide](https://testcontainers.com/guides/getting-started-with-testcontainers-for-nodejs) for testcontainers (minimal repro: https://github.com/FelixZY/bun-testcontainers-stream-issue).
2. Run tests using node and jest (`npm run test:node` in my repro)
3. Run tests using bun (`bun run test:bun` in my repro, or simply `bun test`)

I recommend setting the `DEBUG` envvar to `testcontainers*` for trace logging:

```sh
$ DEBUG=testcontainers* bun test
```

### What is the expected behavior?

Bun should behave identically to node - i.e. the test should run without issue.

### What do you see instead?

Node works fine. In Bun, nothing happens until testcontainers finally times out.

Looking at the difference in log output, it seems Node streams log output to testcontainers in real time, whereas Bun seem to wait for the log stream to close before providing it to testcontainers in a single chunk.

#### Software
* Node: 18.0.0 / 22.7.0
* Docker: 27.2.0, build 3ab4256

#### Packages
* `@testcontainers/postgresql`: 10.13.0
* `jest`: 29.7.0
* `pg`: 8.12.0

#### Logs

<details><summary>Node</summary>

```
DEBUG=testcontainers* npm run test:node

> testcontainer-test@1.0.0 test:node
> jest

  testcontainers [DEBUG] Checking container runtime strategy "TestcontainersHostStrategy"... +0ms
  testcontainers [DEBUG] Container runtime strategy "TestcontainersHostStrategy" is not applicable +0ms
  testcontainers [DEBUG] Checking container runtime strategy "ConfigurationStrategy"... +0ms
  testcontainers [DEBUG] Container runtime strategy "ConfigurationStrategy" is not applicable +0ms
  testcontainers [DEBUG] Checking container runtime strategy "UnixSocketStrategy"... +1ms
  testcontainers [TRACE] Fetching Docker info... +0ms
  testcontainers [TRACE] Fetching remote container runtime socket path... +14ms
  testcontainers [TRACE] Resolving host... +1ms
  testcontainers [TRACE] Fetching Compose info... +0ms
  testcontainers [TRACE] Looking up host IPs... +544ms
  testcontainers [TRACE] Initialising clients... +5ms
  testcontainers [TRACE] Container runtime info:
  testcontainers {
  testcontainers   "node": {
  testcontainers     "version": "v18.0.0",
  testcontainers     "architecture": "x64",
  testcontainers     "platform": "linux"
  testcontainers   },
  testcontainers   "containerRuntime": {
  testcontainers     "host": "localhost",
  testcontainers     "hostIps": [
  testcontainers       {
  testcontainers         "address": "127.0.0.1",
  testcontainers         "family": 4
  testcontainers       }
  testcontainers     ],
  testcontainers     "remoteSocketPath": "/var/run/docker.sock",
  testcontainers     "indexServerAddress": "https://index.docker.io/v1/",
  testcontainers     "serverVersion": "27.2.0",
  testcontainers     "operatingSystem": "Ubuntu 20.04.6 LTS",
  testcontainers     "operatingSystemType": "linux",
  testcontainers     "architecture": "x86_64",
  testcontainers     "cpus": 12,
  testcontainers     "memory": 33579786240,
  testcontainers     "runtimes": [
  testcontainers       "io.containerd.runc.v2",
  testcontainers       "runc"
  testcontainers     ],
  testcontainers     "labels": []
  testcontainers   },
  testcontainers   "compose": {
  testcontainers     "version": "2.4.1",
  testcontainers     "compatability": "v2"
  testcontainers   }
  testcontainers } +1ms
  testcontainers [DEBUG] Container runtime strategy "UnixSocketStrategy" works +1ms
  testcontainers [DEBUG] Checking if image exists "postgres:13.3-alpine"... +2ms
  testcontainers [DEBUG] Checked if image exists "postgres:13.3-alpine" +8ms
  testcontainers [DEBUG] Image "postgres:13.3-alpine" already exists +1ms
  testcontainers [DEBUG] Acquiring lock file "/tmp/testcontainers-node.lock"... +12ms
  testcontainers [DEBUG] Acquired lock file "/tmp/testcontainers-node.lock" +8ms
  testcontainers [DEBUG] Listing containers... +1ms
  testcontainers [DEBUG] Listed containers +3ms
  testcontainers [DEBUG] Creating new Reaper for session "346165467463" with socket path "/var/run/docker.sock"... +2ms
  testcontainers [DEBUG] Checking if image exists "testcontainers/ryuk:0.5.1"... +0ms
  testcontainers [DEBUG] Checked if image exists "testcontainers/ryuk:0.5.1" +2ms
  testcontainers [DEBUG] Image "testcontainers/ryuk:0.5.1" already exists +1ms
  testcontainers [DEBUG] Creating container for image "testcontainers/ryuk:0.5.1"... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Created container for image "testcontainers/ryuk:0.5.1" +96ms
  testcontainers [INFO] [dd49f7998ad6] Starting container for image "testcontainers/ryuk:0.5.1"... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Starting container... +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Started container +258ms
  testcontainers [INFO] [dd49f7998ad6] Started container for image "testcontainers/ryuk:0.5.1" +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Inspecting container... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Inspected container +2ms
  testcontainers [DEBUG] [dd49f7998ad6] Fetching container logs... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Demuxing stream... +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Demuxed stream +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Fetched container logs +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Waiting for container to be ready... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Waiting for log message "/.+ Started!/"... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Fetching container logs... +0ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 Pinging Docker... +0ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 Docker daemon is available! +0ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 Starting on port 8080... +0ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 Started! +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Demuxing stream... +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Demuxed stream +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Fetched container logs +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Log wait strategy complete +2ms
  testcontainers [INFO] [dd49f7998ad6] Container is ready +1ms
  testcontainers [DEBUG] [dd49f7998ad6] Connecting to Reaper (attempt 1) on "localhost:32813"... +0ms
  testcontainers [DEBUG] [dd49f7998ad6] Connected to Reaper +1ms
  testcontainers [DEBUG] Releasing lock file "/tmp/testcontainers-node.lock"... +0ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 New client connected: 172.17.0.1:49310 +5ms
  testcontainers [DEBUG] Released lock file "/tmp/testcontainers-node.lock" +0ms
  testcontainers [DEBUG] Creating container for image "postgres:13.3-alpine"... +1ms
  testcontainers:containers [dd49f7998ad6] 2024/09/02 18:43:29 Adding {"label":{"org.testcontainers.session-id=346165467463":true}} +1ms
  testcontainers [DEBUG] [1b1564d8cf44] Created container for image "postgres:13.3-alpine" +98ms
  testcontainers [INFO] [1b1564d8cf44] Starting container for image "postgres:13.3-alpine"... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Starting container... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Started container +262ms
  testcontainers [INFO] [1b1564d8cf44] Started container for image "postgres:13.3-alpine" +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Inspecting container... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Inspected container +1ms
  testcontainers [DEBUG] [1b1564d8cf44] Fetching container logs... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Demuxing stream... +1ms
  testcontainers [DEBUG] [1b1564d8cf44] Demuxed stream +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Fetched container logs +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Waiting for container to be ready... +1ms
  testcontainers [DEBUG] [1b1564d8cf44] Waiting for log message "/.*database system is ready to accept connections.*/"... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Fetching container logs... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Demuxing stream... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Demuxed stream +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Fetched container logs +0ms
  testcontainers:containers [1b1564d8cf44] The files belonging to this database system will be owned by user "postgres". +365ms
  testcontainers:containers [1b1564d8cf44] This user must also own the server process. +0ms
  testcontainers:containers [1b1564d8cf44]  +1ms
  testcontainers:containers [1b1564d8cf44] The database cluster will be initialized with locale "en_US.utf8". +0ms
  testcontainers:containers [1b1564d8cf44] The default database encoding has accordingly been set to "UTF8". +0ms
  testcontainers:containers [1b1564d8cf44] The default text search configuration will be set to "english". +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] Data page checksums are disabled. +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] fixing permissions on existing directory /var/lib/postgresql/data ... ok +0ms
  testcontainers:containers [1b1564d8cf44] creating subdirectories ... ok +0ms
  testcontainers:containers [1b1564d8cf44] selecting dynamic shared memory implementation ... posix +0ms
  testcontainers:containers [1b1564d8cf44] selecting default max_connections ... 100 +6ms
  testcontainers:containers [1b1564d8cf44] selecting default shared_buffers ... 128MB +11ms
  testcontainers:containers [1b1564d8cf44] selecting default time zone ... UTC +53ms
  testcontainers:containers [1b1564d8cf44] creating configuration files ... ok +2ms
  testcontainers:containers [1b1564d8cf44] running bootstrap script ... ok +130ms
  testcontainers:containers [1b1564d8cf44] sh: locale: not found +168ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:30.180 UTC [30] WARNING:  no usable system locales were found +0ms
  testcontainers:containers [1b1564d8cf44] performing post-bootstrap initialization ... ok +399ms
  testcontainers:containers [1b1564d8cf44] syncing data to disk ... ok +502ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] initdb: warning: enabling "trust" authentication for local connections +0ms
  testcontainers:containers [1b1564d8cf44] You can change this by editing pg_hba.conf or using the option -A, or +0ms
  testcontainers:containers [1b1564d8cf44] --auth-local and --auth-host, the next time you run initdb. +0ms
  testcontainers:containers [1b1564d8cf44]  +1ms
  testcontainers:containers [1b1564d8cf44] Success. You can now start the database server using: +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] pg_ctl -D /var/lib/postgresql/data -l logfile start +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] waiting for server to start....2024-09-02 18:43:31.138 UTC [35] LOG:  starting PostgreSQL 13.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20210424) 10.3.1 20210424, 64-bit +56ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.144 UTC [35] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432" +6ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.166 UTC [36] LOG:  database system was shut down at 2024-09-02 18:43:30 UTC +21ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.176 UTC [35] LOG:  database system is ready to accept connections +11ms
  testcontainers:containers [1b1564d8cf44] done +17ms
  testcontainers:containers [1b1564d8cf44] server started +0ms
  testcontainers:containers [1b1564d8cf44] CREATE DATABASE +171ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/* +1ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] waiting for server to shut down...2024-09-02 18:43:31.366 UTC [35] LOG:  received fast shutdown request +0ms
  testcontainers:containers [1b1564d8cf44] .2024-09-02 18:43:31.372 UTC [35] LOG:  aborting any active transactions +6ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.372 UTC [35] LOG:  background worker "logical replication launcher" (PID 42) exited with exit code 1 +1ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.372 UTC [37] LOG:  shutting down +0ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.423 UTC [35] LOG:  database system is shut down +51ms
  testcontainers:containers [1b1564d8cf44] done +43ms
  testcontainers:containers [1b1564d8cf44] server stopped +0ms
  testcontainers:containers [1b1564d8cf44]  +0ms
  testcontainers:containers [1b1564d8cf44] PostgreSQL init process complete; ready for start up. +0ms
  testcontainers:containers [1b1564d8cf44]  +1ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.509 UTC [1] LOG:  starting PostgreSQL 13.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20210424) 10.3.1 20210424, 64-bit +41ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.509 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432 +1ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.509 UTC [1] LOG:  listening on IPv6 address "::", port 5432 +0ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.520 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432" +11ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.535 UTC [49] LOG:  database system was shut down at 2024-09-02 18:43:31 UTC +15ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.544 UTC [1] LOG:  database system is ready to accept connections +9ms
  testcontainers [DEBUG] [1b1564d8cf44] Log wait strategy complete +2s
  testcontainers [INFO] [1b1564d8cf44] Container is ready +1ms
  testcontainers [INFO] [1b1564d8cf44] Stopping container... +85ms
  testcontainers [DEBUG] [1b1564d8cf44] Stopping container... +0ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.660 UTC [1] LOG:  received fast shutdown request +115ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.666 UTC [1] LOG:  aborting any active transactions +6ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.667 UTC [1] LOG:  background worker "logical replication launcher" (PID 55) exited with exit code 1 +1ms
  testcontainers:containers [1b1564d8cf44] 2024-09-02 18:43:31.667 UTC [50] LOG:  shutting down +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Stopped container +424ms
  testcontainers [DEBUG] [1b1564d8cf44] Removing container... +0ms
  testcontainers [DEBUG] [1b1564d8cf44] Removed container +61ms
  testcontainers [INFO] [1b1564d8cf44] Stopped container +0ms
 PASS  ./customer-repository.test.js
  Customer Repository
    ✓ should create and return multiple customers (27 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        4.119 s
Ran all test suites.
```

</details>

<details><summary>Bun</summary>

```

DEBUG=testcontainers* bun test
bun test v1.1.26 (0a37423b)

customer-repository.test.js:
  testcontainers [DEBUG] Checking container runtime strategy "TestcontainersHostStrategy"... +0ms
  testcontainers [DEBUG] Container runtime strategy "TestcontainersHostStrategy" is not applicable +7ms
  testcontainers [DEBUG] Checking container runtime strategy "ConfigurationStrategy"... +0ms
  testcontainers [DEBUG] Container runtime strategy "ConfigurationStrategy" is not applicable +0ms
  testcontainers [DEBUG] Checking container runtime strategy "UnixSocketStrategy"... +0ms
  testcontainers [TRACE] Fetching Docker info... +1ms
  testcontainers [TRACE] Fetching remote container runtime socket path... +13ms
  testcontainers [TRACE] Resolving host... +0ms
  testcontainers [TRACE] Fetching Compose info... +1ms
  testcontainers [TRACE] Looking up host IPs... +541ms
  testcontainers [TRACE] Initialising clients... +5ms
  testcontainers [TRACE] Container runtime info:
  testcontainers {
  testcontainers   "node": {
  testcontainers     "version": "v22.6.0",
  testcontainers     "architecture": "x64",
  testcontainers     "platform": "linux"
  testcontainers   },
  testcontainers   "containerRuntime": {
  testcontainers     "host": "localhost",
  testcontainers     "hostIps": [
  testcontainers       {
  testcontainers         "address": "127.0.0.1",
  testcontainers         "family": 4
  testcontainers       }
  testcontainers     ],
  testcontainers     "remoteSocketPath": "/var/run/docker.sock",
  testcontainers     "indexServerAddress": "https://index.docker.io/v1/",
  testcontainers     "serverVersion": "27.2.0",
  testcontainers     "operatingSystem": "Ubuntu 20.04.6 LTS",
  testcontainers     "operatingSystemType": "linux",
  testcontainers     "architecture": "x86_64",
  testcontainers     "cpus": 12,
  testcontainers     "memory": 33579786240,
  testcontainers     "runtimes": [
  testcontainers       "io.containerd.runc.v2",
  testcontainers       "runc"
  testcontainers     ],
  testcontainers     "labels": []
  testcontainers   },
  testcontainers   "compose": {
  testcontainers     "version": "2.4.1",
  testcontainers     "compatability": "v2"
  testcontainers   }
  testcontainers } +0ms
  testcontainers [DEBUG] Container runtime strategy "UnixSocketStrategy" works +2ms
  testcontainers [DEBUG] Checking if image exists "postgres:13.3-alpine"... +3ms
  testcontainers [DEBUG] Checked if image exists "postgres:13.3-alpine" +7ms
  testcontainers [DEBUG] Image "postgres:13.3-alpine" already exists +0ms
  testcontainers [DEBUG] Acquiring lock file "/tmp/testcontainers-node.lock"... +14ms
  testcontainers [DEBUG] Acquired lock file "/tmp/testcontainers-node.lock" +5ms
  testcontainers [DEBUG] Listing containers... +1ms
  testcontainers [DEBUG] Listed containers +3ms
  testcontainers [DEBUG] Creating new Reaper for session "a797291752db" with socket path "/var/run/docker.sock"... +1ms
  testcontainers [DEBUG] Checking if image exists "testcontainers/ryuk:0.5.1"... +1ms
  testcontainers [DEBUG] Checked if image exists "testcontainers/ryuk:0.5.1" +4ms
  testcontainers [DEBUG] Image "testcontainers/ryuk:0.5.1" already exists +1ms
  testcontainers [DEBUG] Creating container for image "testcontainers/ryuk:0.5.1"... +1ms
  testcontainers [DEBUG] [5bc13a34f876] Created container for image "testcontainers/ryuk:0.5.1" +195ms
  testcontainers [INFO] [5bc13a34f876] Starting container for image "testcontainers/ryuk:0.5.1"... +0ms
  testcontainers [DEBUG] [5bc13a34f876] Starting container... +1ms
  testcontainers [DEBUG] [5bc13a34f876] Started container +317ms
  testcontainers [INFO] [5bc13a34f876] Started container for image "testcontainers/ryuk:0.5.1" +0ms
  testcontainers [DEBUG] [5bc13a34f876] Inspecting container... +0ms
  testcontainers [DEBUG] [5bc13a34f876] Inspected container +2ms
  testcontainers [DEBUG] [5bc13a34f876] Fetching container logs... +1ms
  testcontainers [DEBUG] [5bc13a34f876] Demuxing stream... +2ms
  testcontainers [DEBUG] [5bc13a34f876] Demuxed stream +0ms
  testcontainers [DEBUG] [5bc13a34f876] Fetched container logs +1ms
  testcontainers [DEBUG] [5bc13a34f876] Waiting for container to be ready... +0ms
  testcontainers [DEBUG] [5bc13a34f876] Waiting for log message "/.+ Started!/"... +0ms
  testcontainers [DEBUG] [5bc13a34f876] Fetching container logs... +0ms
  testcontainers [DEBUG] [5bc13a34f876] Demuxing stream... +2ms
  testcontainers [DEBUG] [5bc13a34f876] Demuxed stream +0ms
  testcontainers [DEBUG] [5bc13a34f876] Fetched container logs +0ms
  testcontainers [ERROR] [5bc13a34f876] Log message "/.+ Started!/" not received after 60000ms +1m
  testcontainers [ERROR] [5bc13a34f876] Container failed to be ready: Error: Log message "/.+ Started!/" not received after 60000ms +1ms
  testcontainers [DEBUG] [5bc13a34f876] Stopping container... +1ms
  testcontainers:containers [5bc13a34f876] 2024/09/02 18:53:28 Pinging Docker... +0ms
  testcontainers:containers [5bc13a34f876] 2024/09/02 18:53:28 Docker daemon is available! +1ms
  testcontainers:containers [5bc13a34f876] 2024/09/02 18:53:28 Starting on port 8080... +0ms
  testcontainers:containers [5bc13a34f876] 2024/09/02 18:53:28 Started! +0ms
  testcontainers:containers [5bc13a34f876] panic: Timed out waiting for the first connection +0ms
  testcontainers:containers [5bc13a34f876]  +1ms
  testcontainers:containers [5bc13a34f876] goroutine 1 [running]: +0ms
  testcontainers:containers [5bc13a34f876] main.waitForPruneCondition({0x846fd0, 0xc0002ca200}, 0xc000094420, 0xc000094480) +0ms
  testcontainers:containers [5bc13a34f876] /app/main.go:229 +0x4d9 +1ms
  testcontainers:containers [5bc13a34f876] main.main() +0ms
  testcontainers:containers [5bc13a34f876] /app/main.go:146 +0x305 +0ms
  testcontainers [DEBUG] [5bc13a34f876] Log wait strategy complete +40ms
  testcontainers [DEBUG] [5bc13a34f876] Stopped container +351ms
  testcontainers [DEBUG] [5bc13a34f876] Removing container... +0ms
  testcontainers [ERROR] [5bc13a34f876] Failed to remove container: Error: (HTTP code 409) unexpected - removal of container 5bc13a34f876500c58495d2bb9b84df00b42094f73901225de4eadf715a71359 is already in progress  +26ms
  testcontainers [ERROR] [5bc13a34f876] Failed to stop container after it failed to be ready: Error: (HTTP code 409) unexpected - removal of container 5bc13a34f876500c58495d2bb9b84df00b42094f73901225de4eadf715a71359 is already in progress  +0ms
  testcontainers [DEBUG] Releasing lock file "/tmp/testcontainers-node.lock"... +0ms
  testcontainers [DEBUG] Released lock file "/tmp/testcontainers-node.lock" +0ms

# Unhandled error between tests
-------------------------------
22 |         const stream = await client.container.logs(container, { since: startTime ? startTime.getTime() / 1000 : 0 });
23 |         return new Promise((resolve, reject) => {
24 |             const timeout = setTimeout(() => {
25 |                 const message = `Log message "${this.message}" not received after ${this.startupTimeout}ms`;
26 |                 common_1.log.error(message, { containerId: container.id });
27 |                 reject(new Error(message));
                            ^
error: Log message "/.+ Started!/" not received after 60000ms
      at /<redacted>/testcontainer-test/node_modules/testcontainers/build/wait-strategies/log-wait-strategy.js:27:24
      at /<redacted>/testcontainer-test/customer-repository.test.js:5:1
-------------------------------

✗ Customer Repository > should create and return multiple customers

 0 pass
 1 fail
 1 error
Ran 1 tests across 1 files. [61.85s]
```

**Comment by FelixZY** [comment]
I've managed to narrow the issue down a bit further. It seems testcontainers is talking http to the docker container via unix domain socket. This is accomplished using the [`docker-modem`](https://github.com/apocas/docker-modem) package. This package creates a (funky?) instance of the native `http` package here: <https://github.com/apocas/docker-modem/blob/master/lib/http.js>. Specifically, these lines concern me:

```js
// Abbreviated!
var h = function() {};
h.prototype = require('http');
h = new h();

h.request = function(h) {
  return function (/*...*/)  {
    /* ... */
    var clientRequest = Object.getPrototypeOf(h).request(options, redirectCallback(reqUrl, redirect));
    /* ... */
    return clientRequest;
  };
}
```

Following the debugger, I'm able to determine that `h.request(opts, function() {}).on('response', (res => /*... */))` does not stream data.

**Comment by FelixZY** [comment]
I looked up the node docs and noted that `on('response', ...)` actually should only be called once; my mistake.

> Emitted when a response is received to this request. This event is emitted only once.

However, the http stream clearly cannot be read in real time. I tested this by adding a breakpoint here:

https://github.com/testcontainers/testcontainers-node/blob/aedb386f81e7749d9336088e9fefd53c454a0353/packages/testcontainers/src/wait-strategies/log-wait-strategy.ts#L36

It seems the listeners are set up immediately as expected but the stream itself does not appear to emit any data before testcontainers time out.

**Comment by FelixZY** [comment]
Me and @assapir exchanged some emails and both noted that there are sporadic cases where Bun actually does pass the test. In my experience, this has been in a small minority of cases (<5%). I also have not found a way of affecting this flakiness, though I theorise it has something to do with running multiple Bun instances or having the target container running before Bun.

**Comment by FelixZY** [comment]
Hey, now this is interesting:

```js
const http = require("http");

const HOST = "localhost";
const PORT = 3000;

const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const server = http.createServer(async (req, res) => {
  res.writeHead(200, {
    "Content-Type": "text/plain",
    "Transfer-Encoding": "chunked",
  });
  await wait(5000);
  for (let i = 0; i < 5; i++) {
    const chunk = `Chunk ${i}\n`;
    res.write(chunk);
    console.log(`${new Date().toISOString()} ==> ${chunk.trimEnd()}`);
    await wait(1000);
  }
  res.end();
});

server.listen(PORT, HOST, async () => {
  await fetch(`http://${HOST}:${PORT}/`)
    .then((response) => response.body.getReader())
    .then(async (reader) => {
      const decoder = new TextDecoder();
      while (true) {
        const { done, value: chunk } = await reader.read();
        if (done) break;
        console.log(`${new Date().toISOString()} <== ${decoder.decode(chunk).trimEnd()}`);
      }
    })
    .then(() => server.unref());
});
```

Node:
```
$ node main.js 
(node:167209) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
2024-09-03T15:02:58.057Z ==> Chunk 0
2024-09-03T15:02:58.065Z <== Chunk 0
2024-09-03T15:02:59.059Z ==> Chunk 1
2024-09-03T15:02:59.060Z <== Chunk 1
2024-09-03T15:03:00.060Z ==> Chunk 2
2024-09-03T15:03:00.062Z <== Chunk 2
2024-09-03T15:03:01.062Z ==> Chunk 3
2024-09-03T15:03:01.063Z <== Chunk 3
2024-09-03T15:03:02.062Z ==> Chunk 4
2024-09-03T15:03:02.063Z <== Chunk 4
```

Bun:

```
$ bun main.js 
2024-09-03T15:03:13.741Z ==> Chunk 0
2024-09-03T15:03:14.744Z ==> Chunk 1
2024-09-03T15:03:15.746Z ==> Chunk 2
2024-09-03T15:03:15.746Z <== Chunk 0
Chunk 1
Chunk 2
2024-09-03T15:03:16.747Z ==> Chunk 3
2024-09-03T15:03:16.748Z <== Chunk 3
2024-09-03T15:03:17.748Z ==> Chunk 4
2024-09-03T15:03:17.748Z <== Chunk 4
```

Looks like there's some sort of delay in Bun when compared to node!

**Comment by FelixZY** [comment]
If I bump the delay between chunks, I still get the same delayed result, indicating it is not an issue of timing. Also, with the time between chunks set to 5s, bun actually crashes before the transfer is complete!

```
$ node main.js
(node:169920) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
2024-09-03T15:07:22.020Z ==> Chunk 0
2024-09-03T15:07:22.026Z <== Chunk 0
2024-09-03T15:07:27.022Z ==> Chunk 1
2024-09-03T15:07:27.024Z <== Chunk 1
2024-09-03T15:07:32.024Z ==> Chunk 2
2024-09-03T15:07:32.026Z <== Chunk 2
2024-09-03T15:07:37.030Z ==> Chunk 3
2024-09-03T15:07:37.030Z <== Chunk 3
2024-09-03T15:07:42.032Z ==> Chunk 4
2024-09-03T15:07:42.034Z <== Chunk 4
```

```
$ bun main.js 
2024-09-03T15:07:00.510Z ==> Chunk 0
2024-09-03T15:07:05.513Z ==> Chunk 1
2024-09-03T15:07:10.515Z ==> Chunk 2
2024-09-03T15:07:10.518Z <== Chunk 0
Chunk 1
Chunk 2
ConnectionClosed: The socket connection was closed unexpectedly. For more information, pass `verbose: true` in the second argument to fetch()
 path: "http://localhost:3000/"

Bun v1.1.26 (Linux x64)
```

**Comment by FelixZY** [comment]
I have investigated this as far as I'm able to without spending (an even more) significant effort. The issue seems to be specifically with Buns implementation of node's `http` module, as `Bun.serve` appears to work flawlessly.

I have [created a PR](https://github.com/oven-sh/bun/pull/13716) with a failing test in the hopes it will be useful for anyone taking the torch from here on.

**Comment by FelixZY** [comment]
Tested with bun 1.1.26, 1.1.20, 1.0.33, 1.0.0. Not a regression.

**Comment by FelixZY** [comment]
I'm just guessing at this point but could `ReadableStreamDefaultReader.readMany` here be the culprit?

https://github.com/oven-sh/bun/blob/a9cf463eebf53dbc2a66cd0d5819f43185d2da63/src/js/node/http.ts#L947

`readMany` is not part of the official `ReadableStreamDefaultReader` spec and is indeed documented as such:

> Only available in Bun. If there are multiple chunks in the queue, this will return all of them at the same time. Will only return a promise if the data is not immediately available.

It looks like `readMany` is defined [here](https://github.com/oven-sh/bun/blob/a9cf463eebf53dbc2a66cd0d5819f43185d2da63/src/js/builtins/ReadableStreamDefaultReader.ts#L46) and a potential issue (again, just guessing!) could be this ("[controller] hasn't been started yet"):

https://github.com/oven-sh/bun/blob/a9cf463eebf53dbc2a66cd0d5819f43185d2da63/src/js/builtins/ReadableStreamDefaultReader.ts#L60-L68

Note that `ReadableStreamDefaultReader.readMany` pushes multiple values without releasing the thread

https://github.com/oven-sh/bun/blob/a9cf463eebf53dbc2a66cd0d5819f43185d2da63/src/js/node/http.ts#L957-L959

 which would seem to correspond with Bun's logging multiple lines at once:
 
 ```
$ bun main.js 
2024-09-03T15:07:00.510Z ==> Chunk 0
2024-09-03T15:07:05.513Z ==> Chunk 1
2024-09-03T15:07:10.515Z ==> Chunk 2
2024-09-03T15:07:10.518Z <== Chunk 0
Chunk 1
Chunk 2
```

**Comment by devcaeg** [comment]
I have a similar problem in `Bun.server` it seems that if you send two chunks immediately, only the first one arrives and the second one is lost forever.

There is also another problem, and that is that it has a “delay” if you send a chunk it is not really sent, you have to send another chunk for the previous one to be sent, that is, there is always a “chunk” delayed.

**Comment by Jarred-Sumner** [comment]
> I have a similar problem in `Bun.server` it seems that if you send two chunks immediately, only the first one arrives and the second one is lost forever.
> 
> There is also another problem, and that is that it has a “delay” if you send a chunk it is not really sent, you have to send another chunk for the previous one to be sent, that is, there is always a “chunk” delayed.

Can you paste some code? The way this currently works is it flushes the data immediately before idling in the event loop, to more efficiently batch up calls to send(). If you use `"type": "direct"` ReadableStream, this behavior would occur if you weren't calling `flush()`. If you're using a regular `ReadableStream` though, this behavior shouldn't occur and it also shouldn't occur if using node:http.

**Comment by devcaeg** [comment]
@Jarred-Sumner 
I made this small example, although I can't reproduce the error that happens in my real application. In a few hours I will try again with another reproduction of the problem.

```ts
export const generator = async function* () {
  yield {
    data: 0,
  };

  yield {
    data: 1,
  };

  let i = 2;

  while (true) {
    yield {
      data: i,
    };

    i++;

    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
};

export const server = Bun.serve({
  fetch: (request) => {
    return new Response(
      new ReadableStream({
        async start(controller) {
          let closed = false as boolean;

          request.signal.addEventListener('abort', () => {
            closed = true;

            try {
              controller.close();
            } catch {}
          });

          for await (const chunk of generator()) {
            if (closed) {
              break;
            }

            controller.enqueue(`data: ${JSON.stringify(chunk.data)}\n\n`);
          }
        },
      }),
      {
        headers: {
          'cache-control': 'no-cache',
          'content-type': 'text/event-stream',
          connection: 'keep-alive',
        },
        status: 200,
      },
    );
  },
  idleTimeout: 0, // TODO: It must be set to 0 due to a bug in the idleTimeout in the latest version of “bun”.
});

console.log(`Server running at ${server.url}`);
```

```ts
import { describe, it, expect } from 'bun:test';

describe('Server test', () => {
  it(
    'ReadableStream Test',
    async () => {
      const response = await fetch('http://localhost:3000');

      expect(response.status).toBe(200);
      expect(response.headers.get('content-type')).toBe('text/event-stream');
      expect(response.body).toBeDefined();

      const reader = response.body?.getReader();

      expect(reader).toBeDefined();

      const decoder = new TextDecoder();

      let i = 1;

      while (true) {
        const { done, value } = await reader!.read();

        if (done || i === 30) {
          break;
        }

        const data = decoder.decode(value);

        expect(data).toContain(`data: ${i}`);

        i++;
      }
    },
    {
      timeout: 0,
    },
  );
});
```

**Comment by Jarred-Sumner** [comment]
@devcaeg yeah this doesn't seem to reproduce the issue

we will need a reproduction to fix this

**Comment by Jarred-Sumner** [comment]
The reproduction in the test is good though

**Comment by FelixZY** [comment]
My own testing seemed to clear `Bun.serve`. @devcaeg are you sure your code is not reaching down to the `http` server behind the scenes? 

@Jarred-Sumner 
> The reproduction in the test is good though

Sorry, would that be "it's good to have a reproduction test", "the reproduction test provided by @devcaeg reproduced the issue" or "the reproduction test in #13716 reproduces the issue"?

**Comment by ptrxyz** [comment]
Seems you guys are much further by now, but I still wanted to confirm the problem with testcontainers.

Thanks to all for digging into this! :+1:

**Comment by GraemeF** [comment]
In the meantime, has anyone found a workaround to get testcontainers working with `bun test`?

**Comment by berenddeboer** [comment]
Also running into this :-(
